File: docs/codeql/codeql-workshops/go/codeql-go-sqli/add-user.go
package main

import (
	"database/sql"
	"fmt"
	"log"
	"os"
	"strings"

	_ "github.com/mattn/go-sqlite3"
)

func get_user_info() (string, error) {
	fmt.Println("Hello, World!")
	fmt.Println("*** Welcome to sql injection ***")
	fmt.Print("Please enter name: ")
	buf := make([]byte, 1024)
	count, err := os.Stdin.Read(buf)
	if err != nil {
		return "", err
	}
	trimmed := strings.TrimSpace(string(buf[0 : count-1]))
	return trimmed, nil
}

func get_new_id() int {
	return os.Getpid()
}

func write_info(id int, info string) {
	db, err := sql.Open("sqlite3", "./users.sqlite")
	if err != nil {
		log.Fatal(err)
	}
	query := fmt.Sprintf("INSERT INTO users VALUES (%d, '%s')", id, info)
	log.Printf("Running query: %s", query)
	_, err = db.Exec(query)
	if err != nil {
		log.Fatal(err)
	}
	db.Close()
}

func main() {
	log.Printf("sqli started")
	info, err := get_user_info()
	if err != nil {
		log.Printf("get_user_info failed: %s", err)
		os.Exit(1)
	}
	id := get_new_id()
	write_info(id, info)
	log.Printf("sqli finished")
}

File: docs/codeql/codeql-workshops/java/codeql-dataflow-sql-injection/AddUser.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;

public class AddUser {
    public static Connection connect() {
        Connection conn = null;
        try {
            String url = "jdbc:sqlite:users.sqlite";
            conn = DriverManager.getConnection(url);
            System.out.println("Connected...");
        } catch (SQLException e) {
            System.out.println(e.getMessage());
        }
        return conn;
    }

    static String get_user_info() {
        System.out.println("Enter name:");
        return System.console().readLine();
    }

    static void write_info(int id, String info) {
        try (Connection conn = connect()) { 
            String query = String.format("INSERT INTO users VALUES (%d, '%s')", id, info);
            conn.createStatement().executeUpdate(query);
            System.err.printf("Sent: %s", query);
        } catch (SQLException e) {
            System.out.println(e.getMessage());
        }
    }
    
    static int get_new_id() {
        return (int)(Math.random()*100000);
    }

    public static void main(String[] args) {
        String info;
        int id;

        info = get_user_info();
        id = get_new_id();
        write_info(id, info);
    }
}
